File: sdk/go/internal/tx/envelope.go
package tx

// google.protobuf.Any encoding
// Fields: type_url(1, string), value(2, bytes)
func encodeAny(typeURL string, value []byte) []byte {
	return ConcatBytes(
		EncodeFieldString(1, typeURL),
		EncodeFieldBytes(2, value),
	)
}

// cosmos.crypto.secp256k1.PubKey encoding
// type_url: "/cosmos.crypto.secp256k1.PubKey"
// Fields: key(1, bytes)
func encodePubKeyAny(pubKey []byte) []byte {
	innerMsg := EncodeFieldBytes(1, pubKey)
	return encodeAny("/cosmos.crypto.secp256k1.PubKey", innerMsg)
}

// ModeInfo for SIGN_MODE_DIRECT
// ModeInfo { single { mode: SIGN_MODE_DIRECT (1) } }
// ModeInfo: field 1 = Single (message)
// Single: field 1 = mode (varint, SIGN_MODE_DIRECT = 1)
func encodeModeInfoDirect() []byte {
	modeField := EncodeFieldVarint(1, 1) // SIGN_MODE_DIRECT = 1
	single := EncodeFieldBytes(1, modeField)
	return single
}

// SignerInfo encoding
// Fields: public_key(1, Any), mode_info(2, ModeInfo), sequence(3, uint64)
func encodeSignerInfo(pubKey []byte, sequence uint64) []byte {
	pubKeyAny := encodePubKeyAny(pubKey)
	modeInfo := encodeModeInfoDirect()

	return ConcatBytes(
		EncodeFieldBytes(1, pubKeyAny),
		EncodeFieldBytes(2, modeInfo),
		EncodeFieldVarint(3, sequence),
	)
}

// Fee encoding
// Fields: amount(1, repeated Coin), gas_limit(2, uint64)
func encodeFee(coins []Coin, gasLimit uint64) []byte {
	parts := make([][]byte, 0, len(coins)+1)
	for _, coin := range coins {
		coinBytes := coin.Encode()
		parts = append(parts, EncodeFieldBytes(1, coinBytes))
	}
	parts = append(parts, EncodeFieldVarint(2, gasLimit))
	return ConcatBytes(parts...)
}

// EncodeTxBody encodes a TxBody.
// Fields: messages(1, repeated Any), memo(2, string), timeout_height(3, uint64)
func EncodeTxBody(messages []MessageEncoder, memo string, timeoutHeight uint64) []byte {
	parts := make([][]byte, 0, len(messages)+2)
	for _, msg := range messages {
		anyBytes := encodeAny(msg.TypeURL(), msg.Encode())
		parts = append(parts, EncodeFieldBytes(1, anyBytes))
	}
	if memo != "" {
		parts = append(parts, EncodeFieldString(2, memo))
	}
	if timeoutHeight > 0 {
		parts = append(parts, EncodeFieldVarint(3, timeoutHeight))
	}
	return ConcatBytes(parts...)
}

// EncodeAuthInfo encodes an AuthInfo.
// Fields: signer_infos(1, repeated SignerInfo), fee(2, Fee)
func EncodeAuthInfo(pubKey []byte, sequence uint64, feeCoins []Coin, gasLimit uint64) []byte {
	signerInfo := encodeSignerInfo(pubKey, sequence)
	fee := encodeFee(feeCoins, gasLimit)

	return ConcatBytes(
		EncodeFieldBytes(1, signerInfo),
		EncodeFieldBytes(2, fee),
	)
}

// EncodeSignDoc encodes a SignDoc for SIGN_MODE_DIRECT.
// Fields: body_bytes(1, bytes), auth_info_bytes(2, bytes), chain_id(3, string), account_number(4, uint64)
func EncodeSignDoc(bodyBytes, authInfoBytes []byte, chainID string, accountNumber uint64) []byte {
	return ConcatBytes(
		EncodeFieldBytes(1, bodyBytes),
		EncodeFieldBytes(2, authInfoBytes),
		EncodeFieldString(3, chainID),
		EncodeFieldVarint(4, accountNumber),
	)
}

// EncodeTxRaw encodes a TxRaw for broadcast.
// Fields: body_bytes(1, bytes), auth_info_bytes(2, bytes), signatures(3, repeated bytes)
func EncodeTxRaw(bodyBytes, authInfoBytes []byte, signatures ...[]byte) []byte {
	parts := [][]byte{
		EncodeFieldBytes(1, bodyBytes),
		EncodeFieldBytes(2, authInfoBytes),
	}
	for _, sig := range signatures {
		parts = append(parts, EncodeFieldBytes(3, sig))
	}
	return ConcatBytes(parts...)
}

File: sdk/csharp/src/Internal/Tx/Envelope.cs
namespace Seocheon.Sdk.Internal.Tx;

/// <summary>
/// Transaction envelope construction: TxBody, AuthInfo, SignDoc, TxRaw.
/// </summary>
public static class Envelope
{
    // === PubKey type URL ===
    private const string PubKeyTypeUrl = "/cosmos.crypto.secp256k1.PubKey";

    /// <summary>
    /// Builds a protobuf Any message wrapping a typed message.
    /// Any: 1=type_url, 2=value
    /// </summary>
    public static byte[] EncodeAny(string typeUrl, byte[] value)
    {
        return Protobuf.Concat(
            Protobuf.EncodeString(1, typeUrl),
            Protobuf.EncodeBytes(2, value)
        );
    }

    /// <summary>
    /// Builds TxBody.
    /// TxBody: 1=messages (repeated Any), 2=memo, 3=timeout_height
    /// </summary>
    public static byte[] EncodeTxBody(byte[] messageAny, string memo = "", ulong timeoutHeight = 0)
    {
        var result = Protobuf.EncodeMessage(1, messageAny);
        if (!string.IsNullOrEmpty(memo))
            result = Protobuf.Concat(result, Protobuf.EncodeString(2, memo));
        if (timeoutHeight > 0)
            result = Protobuf.Concat(result, Protobuf.EncodeUint64(3, timeoutHeight));
        return result;
    }

    /// <summary>
    /// Builds AuthInfo.
    /// AuthInfo: 1=signer_infos (repeated), 2=fee
    /// SignerInfo: 1=public_key (Any), 2=mode_info, 3=sequence
    /// ModeInfo: 1=single (Single)
    /// Single: 1=mode (SIGN_MODE_DIRECT = 1)
    /// Fee: 1=amount (repeated Coin), 2=gas_limit
    /// Coin: 1=denom, 2=amount
    /// </summary>
    public static byte[] EncodeAuthInfo(byte[] pubKey, ulong sequence, ulong gasLimit, ulong feeAmount, string feeDenom)
    {
        // PubKey Any
        var pubKeyValue = Protobuf.EncodeBytes(1, pubKey);
        var pubKeyAny = EncodeAny(PubKeyTypeUrl, pubKeyValue);

        // ModeInfo: Single { mode = SIGN_MODE_DIRECT (1) }
        var single = Protobuf.EncodeUint64(1, 1); // SIGN_MODE_DIRECT
        var modeInfo = Protobuf.EncodeMessage(1, single);

        // SignerInfo
        var signerInfo = Protobuf.Concat(
            Protobuf.EncodeMessage(1, pubKeyAny),
            Protobuf.EncodeMessage(2, modeInfo),
            Protobuf.EncodeUint64(3, sequence)
        );

        // Fee
        var coin = Protobuf.Concat(
            Protobuf.EncodeString(1, feeDenom),
            Protobuf.EncodeString(2, feeAmount.ToString())
        );
        var fee = Protobuf.Concat(
            Protobuf.EncodeMessage(1, coin),
            Protobuf.EncodeUint64(2, gasLimit)
        );

        return Protobuf.Concat(
            Protobuf.EncodeMessage(1, signerInfo),
            Protobuf.EncodeMessage(2, fee)
        );
    }

    /// <summary>
    /// Builds SignDoc.
    /// SignDoc: 1=body_bytes, 2=auth_info_bytes, 3=chain_id, 4=account_number
    /// </summary>
    public static byte[] EncodeSignDoc(byte[] bodyBytes, byte[] authInfoBytes, string chainId, ulong accountNumber)
    {
        return Protobuf.Concat(
            Protobuf.EncodeBytes(1, bodyBytes),
            Protobuf.EncodeBytes(2, authInfoBytes),
            Protobuf.EncodeString(3, chainId),
            Protobuf.EncodeUint64(4, accountNumber)
        );
    }

    /// <summary>
    /// Builds TxRaw.
    /// TxRaw: 1=body_bytes, 2=auth_info_bytes, 3=signatures (repeated)
    /// </summary>
    public static byte[] EncodeTxRaw(byte[] bodyBytes, byte[] authInfoBytes, byte[] signature)
    {
        return Protobuf.Concat(
            Protobuf.EncodeBytes(1, bodyBytes),
            Protobuf.EncodeBytes(2, authInfoBytes),
            Protobuf.EncodeBytes(3, signature)
        );
    }
}
